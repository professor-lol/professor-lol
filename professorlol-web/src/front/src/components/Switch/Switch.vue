<template>
  <label :class="classList">
    <input
      :id="id"
      :checked="isChecked"
      :disabled="disabled"
      :required="required"
      :name="name"
      :value="value"
      class="switch-input form-check-input"
      type="checkbox"
      true-value="value"
      false-value="uncheckedValue"
      @change="handleChange"
    >
    <template v-if="label">
      <span
        :data-checked="dataOn"
        :data-unchecked="dataOff"
        class="switch-slider">
      </span>
    </template>
    <template v-else>
      <span class="switch-slider"></span>
    </template>
  </label>
</template>

<script>
export default {
  model: {
    prop: 'modelChecked',
    event: 'change'
  },
  props: {
    id: {
      type: String,
      default: function () {
        return 'switch-id-' + this._uid;
      },
    },
    color: {
      type: String,
      default: 'secondary'
    },
    label: {
      type: Boolean,
      default: null
    },
    outline: {
      type: [Boolean, String],
      default: null,
      validator: value => [false, true, '', 'alt'].indexOf(value) !== -1
    },
    size: {
      type: String,
      default: null,
      validator: value => ['', 'lg', 'sm'].indexOf(value) !== -1
    },
    checked: {
      type: Boolean,
      default: false
    },
    defaultChecked: {
      type: Boolean,
      default: false
    },
    modelChecked: {
      default: undefined
    },
    value: {
      default: true
    },
    disabled: {
      type: Boolean,
      default: false
    },
    name: {
      type: String,
      default: null
    },
    required: {
      type: Boolean,
      default: false
    },
    onChange: {
      type: Function
    },
    uncheckedValue: {
      default: false
    },
    variant: {
      type: String,
      default: null,
      validator: value => [null, '3d', 'pill'].indexOf(value) !== -1
    },
    dataOn: {
      type: String,
      default: 'On'
    },
    dataOff: {
      type: String,
      default: 'Off'
    }
  },
  data: function () {
    return {
      checkedData: this.isChecked ? this.value : this.uncheckedValue
    }
  },
  computed: {
    classList () {
      return [
        'switch',
        this.label ? 'switch-label' : '',
        this.size ? `switch-${this.size}` : '',
        this.variant ? `switch-${this.variant}` : '',
        `switch${this.outline ? '-outline' : ''}-${this.color}${this.outline==='alt' ? '-alt' : ''}`,
        'form-check-label'
      ]
    },
    isChecked () {
      if (this.modelChecked === undefined ) {
        return this.checkedData === this.value
      }
      return this.modelChecked === this.value
    }
  },
  mounted() {
    this.toggle(this.defaultChecked || this.checked || this.isChecked);
  },
  methods: {
    handleChange (event) {
      this.toggle(event.target.checked);
    },
    toggle (checked) {
      this.checkedData = checked ? this.value : this.uncheckedValue
      this.$emit('change', this.checkedData);
    }
  },
}
</script>
