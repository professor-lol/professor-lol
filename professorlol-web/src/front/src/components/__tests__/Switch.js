import { mount } from "@vue/test-utils";
import Switch from "../Switch/Switch";

describe("Switch.vue", () => {
  // Inspect the raw component options
  it("should have default props", () => {
    const wrapper = mount(Switch, {
      propsData: {
        color: 'info',
        size: 'lg',
        checked: false,
        defaultChecked: false,
        value: true,
        disabled: false,
        required: false,
        uncheckedValue: false,
        dataOn: 'On',
        dataOff: 'Off',
      }
    })
    expect(wrapper.props().color).toBe('info')
    expect(wrapper.props().size).toEqual('lg')
    expect(wrapper.props().checked).toBe(false)
    expect(wrapper.props().defaultChecked).toBe(false)
    expect(wrapper.props().value).toBe(true)
    expect(wrapper.props().disabled).toBe(false)
    expect(wrapper.props().required).toBe(false)
    expect(wrapper.props().dataOn).toBe('On')
    expect(wrapper.props().dataOff).toBe('Off')
  });
  it('has computed properties', () => {
    expect(typeof Switch.computed.classList).toBe('function')
    expect(typeof Switch.computed.isChecked).toBe('function')
  })
  it('has a handleChange method', () => {
    expect(typeof Switch.methods.handleChange).toBe('function')
  })
  it('has a toggle method', () => {
    expect(typeof Switch.methods.toggle).toBe('function')
  })
  it('renders correctly', () => {
    const wrapper = mount(Switch)
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.is('label')).toBe(true)
    expect(wrapper.classes()).toContain('switch')
    expect(wrapper.find('input').classes()).toContain('switch-input')
    expect(wrapper.find('span').classes()).toContain('switch-slider')
  })
});
