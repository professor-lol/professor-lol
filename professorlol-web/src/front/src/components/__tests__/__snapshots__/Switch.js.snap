// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch.vue renders correctly 1`] = `
<label
  class="switch switch-secondary form-check-label"
>
  <input
    class="switch-input form-check-input"
    false-value="uncheckedValue"
    id="switch-id-3"
    true-value="value"
    type="checkbox"
    value="true"
  />
   
  <span
    class="switch-slider"
  />
</label>
`;
